File: nvcs/pattern.py
'''
Created on Mar 15, 2013

@author: knimerfro
'''

class Pattern:
    attrpatts = dict()
    
    @classmethod
    def register(cls, attr, pattcls):
        Pattern.attrpatts[attr] = pattcls
        
    def __init__(self, expr):
        if (not expr):
            raise RuntimeError("Pattern is empty")
        else:
            self.negated = expr.lstrip().startswith('!')
            self.pattern = [s.strip() for s in expr.lstrip(' !').split(',')]
        
    def match(self, value):
        #print("match:", self, value)
        if (self.negated):
            #print("   result=", not value in self.pattern)
            return not value in self.pattern
        else:
            #print("   result=", value in self.pattern)
            return value in self.pattern
            
    def __repr__(self):
        return "Pattern(%r, %r)" % ('!' if self.negated else ' ', self.pattern)

class PatternGroup:
    def __init__(self, pattgrp):
        self.pattgrp = dict()
        for attr in pattgrp.keys():
            self.pattgrp[attr] = Pattern.attrpatts[attr](pattgrp[attr])
    
    def match_all(self, valugrp):
        result = True
        for attr in self.pattgrp.keys():
            #print("match_all:", attr, self.pattgrp[attr], valugrp[attr])
            if (not self.pattgrp[attr].match(valugrp[attr])):
                result = False
                break
        return result

    def __repr__(self):
        return "PatternGroup(%r)" % (self.pattgrp)

class PatternList:
    def __init__(self, label, *pattgrps):
        if (not pattgrps):
            raise RuntimeError("PatternList is empty")
        self.label = label
        self.alternatives = [PatternGroup(pattgrp) for pattgrp in pattgrps]
        
    def match_any(self, valugrp):
        #print("match_any:", valugrp)
        if (not self.alternatives):
            result = True
        else:
            result = False
            for patterngroup in self.alternatives:
                if (patterngroup.match_all(valugrp)):
                    result = True
                    break
        return result
    
    def __repr__(self):
        return "PatternList(%r, %r)" % (self.label, self.alternatives)

File: nvcs_java/src/main/java/nvcs_patterns/Pattern.java
package nvcs_patterns;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Pattern {

    public static Map<String,String> attrpatts = new HashMap<>();
    public List<String> pattern;
    public boolean negated;
    
    public static void register(String attr, String pattcls) {
        attrpatts.put(attr, pattcls);
    }

    public Pattern(String expr) {
        if (expr == null) {
            throw new RuntimeException("Pattern is empty");
        }
        else {
            String cleanedExpr = expr.replaceAll("^\\s+", "");
            negated = cleanedExpr.startsWith("!");
            pattern = new ArrayList<>();
            for (String splitExpr : cleanedExpr.replace("!", "").split(",")) {
                String stripped = splitExpr.trim();
                pattern.add(stripped);
            }
        }
    }

    public boolean match(Object value) {
        if (negated) {
            return !pattern.contains(value);
        }
        else {
            return pattern.contains(value);
        }
    }

    @Override
    public String toString() {
        return String.format("Pattern(%s, %s)", negated ? "!" : " ", pattern);
    }

}
